prov/rxm/src/rxm_ep.c

```diff
@@ -2121,6 +2121,13 @@ static void rxm_ep_settings_init(struct rxm_ep *rxm_ep)
 	rxm_ep->rxm_mr_local = ofi_mr_local(rxm_ep->rxm_info);
 
 	rxm_ep->inject_limit = rxm_ep->rxm_info->tx_attr->inject_size;
+	/* Adjust RxM inject limit that is used when compraring with user data length.
+	 * This value = 0 if the maximum supported inject limit = 0. Otherwise, this
+	 * value = (maximum supported inject limit + 1).
+	 * This allows to use full inject range and ensure that MSG `fi_inject` won't
+	 * be used in case of the maximum supported inject limit = 0. */
+	rxm_ep->inject_upper_limit = (rxm_ep->inject_limit ?
+				      (rxm_ep->inject_limit + 1) : 0);
```

Comment:
I'm really confused now.  inject_limit is already the maximum inject size + 1.  Now we define a new inject_upper_limit as inject_limit + 1?  These checks aren't making any sense.

Reply:
> inject_limit is already the maximum inject size + 1.

The `inject_limit` is a `fi_info::tx_attr::inject_size` value reported by MSG provider.
There are two cases:
1) `inject_upper_limit` = `fi_info::tx_attr::inject_size` - `sizeof(rxm_pkt)` + `1` = `inject_limit` + `1`
 if `fi_info::tx_attr::inject_size` > `sizeof(rxm_pkt)`
2) `inject_upper_limit` = `0`
if `fi_info::tx_attr::inject_size` <= `sizeof(rxm_pkt)`

The `inject_limit` is the value that RxM reports to applications and uses for inject buffer allocation, but `inject_upper_limit` is internal value that RxM uses to compare with passed data length by user. We use "<" in those comparisons to ensure that full Inject range is used, i.e. [0, `fi_info::tx_attr::inject_size`]. Also this comparison handles case when MSG provider reports `fi_info::tx_attr::inject_size` < `sizeof(rxm_pkt)` (as i40iw and hfi1 verbs drivers do) - we don't use MSG provider's `fi_inject()` in this case.

I don't know easier approach how we can handle these cases without adding overhead (additional checks or addition (`inject_limit + 1`)) in the critical path.

BTW, it seems to be a lack of additional capability `FI_INJECT` that could show whether provder support MSG/TAGGED/RMA operations or not. Because, as @a-ilango correctly noted:

> Hmm. Wouldn't this disallow using fi_inject for sending 0 byte messages but with CQ data?

When Verbs reports `fi_info::tx_attr::inject_size` = 0, it doesn't support inject (i.e. Verbs Inline) transfers at all (even for 0 bytes).
But for other provider it can be interpreted that applications can use `fi_inject` for 0-byte messages. For example, such provider is RxM, there are two cases for `inject_size = 0`:
1) RxM doesn't support `fi_inject` at all - core provider has `inject_size` < `sizeof(rxm_pkt)` (= 64). For example, the core provider reports `inject_size = 63`, RxM's header doesn't fit into this buffer, so RxM reports `inject_size = 0` and an application must not use `fi_[t]inject[data]`.
2) RxM supports `fi_[t]inject[data]` for 0-byte messages - core provider has `inject_size` >= `sizeof(rxm_pkt)` (= 64). For example, the core provider reports `inject_size = 64`, RxM's header fits into this buffer, so RxM reports `inject_size = MSG's inejct_size - sizeof(rxm_pkt) = 0` and an application can use `fi_[t]inject[data]` to send 0-byte messages with/without `tag` and `with/without` data.

Since, we don't have any mechanism to report to user what `inejct_size = 0` means, I think we can consider consider that the application/RxM can't use `fi_[t]inject[data]` in case of some provider reports `inject_size = 0`.

Reply:
RxM won't be able to support 0 length injects if core provider inject size is less than RxM header. In such cases the only option is to return error for 0 length fi_inject and fi_sendmsg with (FI_INJECT | ^FI_COMPLETION).

> I don't know easier approach how we can handle these cases without adding overhead (additional checks or addition (inject_limit + 1)) in the critical path.

So the question is do we add extra check for len == 0 or have an extra limit: inject_upper_limit which avoids this check by modifying the comparison to '<'. @dmitrygladkov is that right?

My vote would be for adding the extra check if it doesn't impact performance and re-visit inject_upper_limit after 1.7.

Reply:
> Since there's no actual buffer when size = 0, it just requires hiding the completion. CQ data shouldn't impact this.

I see. Thanks for the explanation.

> I think the problem here is that you are wanting providers to report that inject isn't supported at all, which isn't supported by the API.

Yes, we should handle it somehow for RxM provider over i40iw Verbs (inline size = 48 by default).
I have a patch that simplifies RxM header and reduces its size (it fixes support of i40iw). RxM should raise a warning if underlying provider doesn't support inject size >= RxM header (or maybe fail initialization RxM over such providers)?
What do you think of this approach?

Or do you want to implement capability in the 1.7 that should notify application about unsupported `FI_INJECT`?

Reply:
How will rxm handle 0 length sends?  RxM has to handle messages larger than the inject size, so it's header already needs to be registered memory.  What difference does it make if the app wants to send 0 bytes versus a million?  If the underlying provider has an inject size that's less than the rxm header, just fall back to the send path.

For verbs, if inline is supported > 0, set its inject function to a fast path that uses it.  If inline isn't supported, set inject to an optimized path that only supports 0 byte transfers.  Have rxm do something similar.

Reply:
> How will rxm handle 0 length sends? RxM has to handle messages larger than the inject size, so it's header already needs to be registered memory. What difference does it make if the app wants to send 0 bytes versus a million? If the underlying provider has an inject size that's less than the rxm header, just fall back to the send path.

The problem with falling back to send path is if there's an error (e.g. other than ENOMEM) on sending the message, we would get to know only when reading the send completion but we can't report the error to app because from app's perspective inject has succeeded and it won't read the completion queue. Handling 0 length sends should be fine as app would read the completion queue to check the status.

Reply:
Even an inject operation can report an error.  The return code from the inject function just means that the operation was queued for processing.

Reply:
> Even an inject operation can report an error. The return code from the inject function just means that the operation was queued for processing.

So app should still check the completion queue following fi_inject?

Also, since fi_inject doesn't accept a context parameter, reporting the correct op_context in error entry is an issue. Apps check the op_context for any error completion.

Reply:
Operations that complete in error always generate a CQ entry.  If no op_context was provided, it should be NULL, and then it's up to the app to figure out what, exacty, failed, since it didn't want any completions.  The inject call should match the behavior of FI_INJECT | ^FI_SELECTIVE_COMPLETION, as you mentioned.
```
If FI_SELECTIVE_COMPLETION is specified, data transfer operations will not generate entries for successful completions
```
This is true with verbs inline data as well.  A completion is still generated.  If we need to document reserved values for op_context (say 0xfff..fff - 0xfff..f00) we can do that.  I don't think this will break anything.